pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/WorkerUtils.java

```diff
@@ -343,4 +349,47 @@ public static boolean isFunctionCodeBuiltin(Function.FunctionDetailsOrBuilder fu
                 .startMessageId(startMessageId)
                 .create();
     }
+
+    public static Producer<byte[]> createExclusiveProducerWithRetry(PulsarClient client,
+                                                                    String topic,
+                                                                    String producerName) {
+        Actions.Action createProducerAction = Actions.Action.builder()
+                .actionName(String.format("Creating exclusive producer for topic %s", topic))
+                .numRetries(5)
+                .sleepBetweenInvocationsMs(10000)
+                .supplier(() -> {
+                    try {
+                        Producer<byte[]> producer = client.newProducer().topic(topic)
+                                .accessMode(ProducerAccessMode.Exclusive)
```

Comment:
Would it make sense to use `WaitForExclusive` here? So that the producer instance will stay pending until it becomes the leader?

Reply:
I don't think it is a good idea for "liveness" is we block forever which may cause deadlocks.  While blocking, the worker might be not even be the leader anymore.  I think we should keep retrying and periodically check if we are still the leader.

Reply:
I agree with @jerrypeng